kong/clustering/rpc/concentrator.lua

```diff
@@ -0,0 +1,299 @@
+    manager = manager,
+    db = db,
+    interest = {}, -- id: callback pair
+    sub_unsub = queue.new(4096), -- pub/sub event queue, executed on the read thread
+    sequence = 0,
+  }
+
+  return setmetatable(self, _MT)
+end
+
+
+function _M:_get_next_id()
+  local res = self.sequence
+  self.sequence = res + 1
+
+  return res
+end
+
+
+local function enqueue_notifications(notifications, notifications_queue)
+  assert(notifications_queue)
+
+  if notifications then
+    for _, n in ipairs(notifications) do
+      assert(notifications_queue:push(n))
+    end
+  end
+end
+
+
+function _M:_event_loop(lconn)
+  local notifications_queue = queue.new(4096)
+  local rpc_resp_channel_name = RESP_CHANNEL_PREFIX .. self.worker_id
+
+  -- we always subscribe to our worker's receiving channel first
+  local res, err = lconn:query('LISTEN "' .. rpc_resp_channel_name .. '";')
+  if not res then
+    return nil, "unable to subscribe to concentrator response channel: " .. err
+  end
+
+  while not exiting() do
+    repeat
+      local n, err = notifications_queue:pop(0)
+      if err then
+        return nil, "unable to pop from notifications queue: " .. err
+      end
+
+      if n then
+        assert(n.operation == "notification")
+
+        if n.channel == rpc_resp_channel_name then
+          -- an response for a previous RPC call we asked for
+          local payload = assert(cjson_decode(n.payload))
+          assert(payload.jsonrpc == "2.0")
+
+          -- response
+          local cb = self.interest[payload.id]
+          self.interest[payload.id] = nil -- edge trigger only once
+
+          if cb then
+            local res, err = cb(payload)
+            if not res then
+              ngx_log(ngx_WARN, "[rpc] concentrator response interest handler failed: id: ",
+                      payload.id, ", err: ", err)
+            end
+
+          else
+            ngx_log(ngx_WARN, "[rpc] no interest for concentrator response id: ", payload.id, ", dropping it")
+          end
+
+        else
+          -- other CP inside the cluster asked us to forward a call
+          assert(n.channel:sub(1, #REQ_CHANNEL_PREFIX) == REQ_CHANNEL_PREFIX)
+
+          local target_id = n.channel:sub(9)
+          local sql = string_format(RPC_REQUEST_DEQUEUE_SQL, self.db.connector:escape_literal(target_id))
+          local calls, err = self.db.connector:query(sql)
+          if not calls then
+            return nil, "concentrator request dequeue query failed: " .. err
+          end
+
+          assert(calls[1] == true)
+          ngx_log(ngx_DEBUG, "concentrator got ", calls[2].affected_rows,
+                  " calls from database for node ", target_id)
+          for _, call in ipairs(calls[2]) do
+            local payload = assert(cjson_decode(call.payload))
+            local reply_to = assert(call.reply_to,
+                                    "unknown requester for RPC")
+
+            local res, err = self.manager:_call(target_id, payload.method,
```

Comment:
I think `concentrator` calling a _"private"_ method on `manager` is not ideal. If that method is used outside of manager itself it shouldn't be prefixed with underscore otherwise it might cause bugs with refactoring later on.

Reply:
Everything within the `rpc` directory is part of the same bundle, and `_` is signaling code outside of `kong.clustering.rpc` should not call these methods directly.

This is a very common pattern, for example, [Rust also have the concept of "crate level visibility"](https://doc.rust-lang.org/reference/visibility-and-privacy.html) which is similiar to the ideas here.